### ISSUE TITLE
Untraceable build error

### ISSUE THREAD

**Issue by jordanpurinton** [ionic-team/ionic#13048]
**Ionic version:**  (check one with "x")
[ ] **2.x**
[x ] **3.x**
[ ] **4.x**

**I'm submitting a ...**  (check one with "x")
[x ] bug report
[ ] feature request

**Current behavior:**
When I try to build for Android I keep getting a simple error that appears as the following:
```
C:\path\to\project>ionic cordova run android --verbose --debug
[DEBUG] CLI flags: { interactive: true, confirm: false }
[DEBUG] Daemon found (pid: 312)
[INFO] Running app-scripts build: --platform android --target cordova

[12:58:34]  build dev started ...
[12:58:34]  clean started ...
[12:58:34]  clean finished in 15 ms
[12:58:34]  copy started ...
[12:58:34]  transpile started ...
[12:58:39]  transpile finished in 5.02 s
[12:58:39]  preprocess started ...
[12:58:39]  deeplinks started ...
[12:58:39]  deeplinks finished in 200 ms
[12:58:39]  preprocess finished in 202 ms
[12:58:39]  webpack started ...
[12:58:40]  copy finished in 5.54 s
[12:58:50]  webpack finished in 10.64 s
[12:58:50]  sass started ...
[12:58:51]  sass finished in 1.11 s
[12:58:51]  postprocess started ...
[12:58:51]  removed unused font files
[12:58:51]  postprocess finished in 34 ms
[12:58:51]  lint started ...
[12:58:51]  build dev finished in 17.24 s
> ionic cordova prepare android
> cordova prepare android
Error: Cannot read property 'length' of undefined

[ERROR] An error occurred while running cordova prepare android (exit code 1).
```
**Expected behavior:**
I've tried removing and re-adding the android platform, I've tried switching npm versions, I've tried deleting `node_modules` and several other things and I still haven't been able to successfully build my project. Does anyone know how I could go about resolving this issue?

**Ionic info:**
```
cli packages: (C:\path\to\project\node_modules)

    @ionic/cli-utils  : 1.12.0
    ionic (Ionic CLI) : 3.12.0

global packages:

    cordova (Cordova CLI) : 7.0.1

local packages:

    @ionic/app-scripts : 2.1.4
    Cordova Platforms  : android 6.2.3
    Ionic Framework    : ionic-angular 3.7.1

System:

    Android SDK Tools : 25.2.5
    Node              : v8.6.0
    npm               : 5.3.0
    OS                : Windows 10

Misc:

    backend : legacy


```

**Comment by kensodemann** [comment]
Hello! Thank you for opening an issue with us!

What do you get if you run `cordova prepare android --verbose`?  (after running your original command to failure)

Thank you for using Ionic

**Comment by jordanpurinton** [comment]
I ran that command and still got `Error: Cannot read property 'length' of undefined`

**Comment by kensodemann** [comment]
I expected you would still get the error, but with the `--verbose` flag I was also hoping for a little more context. There wasn't any?

**Comment by jordanpurinton** [comment]
Here's the last few lines of output

`Updating icons at platforms\android\res
Updating splash screens at platforms\android\res
This app does not have additional resource files defined
Prepared android project successfully
Executing script found in plugin cordova-android-play-services-gradle-release for hook "after_prepare": plugins\cordova-android-play-services-gradle-release\scripts\after-prepare.js
Executing script found in plugin cordova-android-support-gradle-release for hook "after_prepare": plugins\cordova-android-support-gradle-release\scripts\after-prepare.js
Error: Cannot read property 'length' of undefined`

**Comment by kensodemann** [comment]
I am going to hope there is some sanity here and say that the error is probably coming from `plugins\cordova-android-support-gradle-release\scripts\after-prepare.js`.
 
One thing I just noticed is this: `@ionic/app-scripts : 2.1.4`

A couple of things to try:

1. upgrade @ionic/app-scripts to `3.0.0`
2. remove the plugins directory and then do an `ionic build` (may fail first time, if so try just the `ionic build` bit again)
3. if all else fails, try duplicating via a minimal sample app and put that out on GitHub so we can try it out (I build my test apps for Android all the time and haven't duplicated this exact error, but perhaps we have slightly different setups to our apps).

**Comment by jordanpurinton** [comment]
Could I perhaps post the contents of my package.json and config.xml? I only say that because I'm 99 percent sure this is a package/dependency issue.

**Comment by kensodemann** [comment]
yup, go for it

**Comment by jordanpurinton** [comment]
<content src="index.html" />
    <access origin="*" />
    <allow-navigation href="http://ionic.local/*" />
    <allow-navigation href="http://*/*" />
    <allow-navigation href="https://*/*" />
    <allow-navigation href="data:*" />
    <allow-navigation href="*" />
    <allow-navigation href="http://192.168.2.37:8100" />
    <allow-navigation href="http://192.168.56.1:8100" />
    <allow-navigation href="http://192.168.56.1:8101" />
    <allow-intent href="http://*/*" />
    <allow-intent href="https://*/*" />
    <allow-intent href="tel:*" />
    <allow-intent href="sms:*" />
    <allow-intent href="mailto:*" />
    <allow-intent href="geo:*" />
    <allow-intent href="itms-apps:*" />
    <platform name="android">
        <allow-intent href="market:*" />
        <icon density="ldpi" src="resources/android/icon/drawable-ldpi-icon.png" />
        <icon density="mdpi" src="resources/android/icon/drawable-mdpi-icon.png" />
        <icon density="hdpi" src="resources/android/icon/drawable-hdpi-icon.png" />
        <icon density="xhdpi" src="resources/android/icon/drawable-xhdpi-icon.png" />
        <icon density="xxhdpi" src="resources/android/icon/drawable-xxhdpi-icon.png" />
        <icon density="xxxhdpi" src="resources/android/icon/drawable-xxxhdpi-icon.png" />
        <splash density="land-ldpi" src="resources/android/splash/drawable-land-ldpi-screen.png" />
        <splash density="land-mdpi" src="resources/android/splash/drawable-land-mdpi-screen.png" />
        <splash density="land-hdpi" src="resources/android/splash/drawable-land-hdpi-screen.png" />
        <splash density="land-xhdpi" src="resources/android/splash/drawable-land-xhdpi-screen.png" />
        <splash density="land-xxhdpi" src="resources/android/splash/drawable-land-xxhdpi-screen.png" />
        <splash density="land-xxxhdpi" src="resources/android/splash/drawable-land-xxxhdpi-screen.png" />
        <splash density="port-ldpi" src="resources/android/splash/drawable-port-ldpi-screen.png" />
        <splash density="port-mdpi" src="resources/android/splash/drawable-port-mdpi-screen.png" />
        <splash density="port-hdpi" src="resources/android/splash/drawable-port-hdpi-screen.png" />
        <splash density="port-xhdpi" src="resources/android/splash/drawable-port-xhdpi-screen.png" />
        <splash density="port-xxhdpi" src="resources/android/splash/drawable-port-xxhdpi-screen.png" />
        <splash density="port-xxxhdpi" src="resources/android/splash/drawable-port-xxxhdpi-screen.png" />
    </platform>
    <platform name="ios">
        <icon height="57" src="resources/ios/icon/icon.png" width="57" />
        <icon height="114" src="resources/ios/icon/icon@2x.png" width="114" />
        <icon height="40" src="resources/ios/icon/icon-40.png" width="40" />
        <icon height="80" src="resources/ios/icon/icon-40@2x.png" width="80" />
        <icon height="120" src="resources/ios/icon/icon-40@3x.png" width="120" />
        <icon height="50" src="resources/ios/icon/icon-50.png" width="50" />
        <icon height="100" src="resources/ios/icon/icon-50@2x.png" width="100" />
        <icon height="60" src="resources/ios/icon/icon-60.png" width="60" />
        <icon height="120" src="resources/ios/icon/icon-60@2x.png" width="120" />
        <icon height="180" src="resources/ios/icon/icon-60@3x.png" width="180" />
        <icon height="72" src="resources/ios/icon/icon-72.png" width="72" />
        <icon height="144" src="resources/ios/icon/icon-72@2x.png" width="144" />
        <icon height="76" src="resources/ios/icon/icon-76.png" width="76" />
        <icon height="152" src="resources/ios/icon/icon-76@2x.png" width="152" />
        <icon height="167" src="resources/ios/icon/icon-83.5@2x.png" width="167" />
        <icon height="29" src="resources/ios/icon/icon-small.png" width="29" />
        <icon height="58" src="resources/ios/icon/icon-small@2x.png" width="58" />
        <icon height="87" src="resources/ios/icon/icon-small@3x.png" width="87" />
        <splash height="1136" src="resources/ios/splash/Default-568h@2x~iphone.png" width="640" />
        <splash height="1334" src="resources/ios/splash/Default-667h.png" width="750" />
        <splash height="2208" src="resources/ios/splash/Default-736h.png" width="1242" />
        <splash height="1242" src="resources/ios/splash/Default-Landscape-736h.png" width="2208" />
        <splash height="1536" src="resources/ios/splash/Default-Landscape@2x~ipad.png" width="2048" />
        <splash height="2048" src="resources/ios/splash/Default-Landscape@~ipadpro.png" width="2732" />
        <splash height="768" src="resources/ios/splash/Default-Landscape~ipad.png" width="1024" />
        <splash height="2048" src="resources/ios/splash/Default-Portrait@2x~ipad.png" width="1536" />
        <splash height="2732" src="resources/ios/splash/Default-Portrait@~ipadpro.png" width="2048" />
        <splash height="1024" src="resources/ios/splash/Default-Portrait~ipad.png" width="768" />
        <splash height="960" src="resources/ios/splash/Default@2x~iphone.png" width="640" />
        <splash height="480" src="resources/ios/splash/Default~iphone.png" width="320" />
        <allow-intent href="itms:*" />
        <allow-intent href="itms-apps:*" />
    </platform>
    <preference name="webviewbounce" value="true" />
    <preference name="UIWebViewBounce" value="true" />
    <preference name="DisallowOverscroll" value="false" />
    <preference name="android-minSdkVersion" value="16" />
    <preference name="BackupWebStorage" value="none" />
    <preference name="CrosswalkAnimatable" value="false" />
    <preference name="loadUrlTimeoutValue" value="700000" />
    <preference name="android-targetSdkVersion" value="23" />
    <preference name="Fullscreen" value="false" />
    <preference name="SplashScreen" value="screen" />
    <preference name="AutoHideSplashScreen" value="true" />
    <preference name="ShowSplashScreenSpinner" value="false" />
    <preference name="SplashScreenDelay" value="30000" />
    <preference name="FadeSplashScreen" value="true" />
    <preference name="FadeSplashScreenDuration" value="2000" />
    <preference name="android-windowSoftInputMode" value="stateVisible|adjustResize" />
    <feature name="SplashScreen">
        <param name="android-package" value="org.apache.cordova.splashscreen.SplashScreen" />
    </feature>
    <feature name="StatusBar">
        <param name="ios-package" onload="true" value="CDVDevice" />
    </feature>
    <allow-navigation href="http://192.168.2.37:8101" />
    <preference name="SplashShowOnlyFirstTime" value="false" />
    <allow-navigation href="http://192.168.2.37:8102" />
    <engine name="android" spec="^6.3.0" />
    <engine name="ios" spec="^4.4.0" />
    <plugin name="cordova-android-play-services-gradle-release" spec="^1.1.1" />
    <plugin name="cordova-android-support-gradle-release" spec="^1.1.1">
        <variable name="ANDROID_SUPPORT_VERSION" value="25.+" />
    </plugin>
    <plugin name="cordova-open-native-settings" spec="^1.4.1" />
    <plugin name="cordova-plugin-app-version" spec="^0.1.9" />
    <plugin name="cordova-plugin-compat" spec="^1.1.0" />
    <plugin name="cordova-plugin-crosswalk-certificate" spec="^2.0.1" />
    <plugin name="cordova-plugin-crosswalk-webview" spec="^2.3.0">
        <variable name="XWALK_VERSION" value="23+" />
        <variable name="XWALK_LITEVERSION" value="xwalk_core_library_canary:17+" />
        <variable name="XWALK_COMMANDLINE" value="--disable-pull-to-refresh-effect" />
        <variable name="XWALK_MODE" value="embedded" />
        <variable name="XWALK_MULTIPLEAPK" value="true" />
    </plugin>
    <plugin name="cordova-plugin-device" spec="^1.1.6" />
    <plugin name="cordova-plugin-file" spec="^4.3.3" />
    <plugin name="cordova-plugin-file-transfer" spec="^1.6.3" />
    <plugin name="cordova-plugin-filepath" spec="^1.0.2" />
    <plugin name="cordova-plugin-geolocation" spec="^2.4.3">
        <variable name="GEOLOCATION_USAGE_DESCRIPTION" value=" " />
    </plugin>
    <plugin name="cordova-plugin-inappbrowser" spec="^1.7.1" />
    <plugin name="cordova-plugin-splashscreen" spec="^4.0.3" />
    <plugin name="cordova-plugin-statusbar" spec="^2.2.3" />
    <plugin name="cordova-plugin-whitelist" spec="^1.3.2" />
    <plugin name="cordova.plugins.diagnostic" spec="^3.6.7" />
    <plugin name="ionic-plugin-keyboard" spec="^2.2.1" />
</widget>

**Comment by jordanpurinton** [comment]
```
package.json

 "dependencies": {
        "@angular/common": "4.4.4",
        "@angular/compiler": "4.4.4",
        "@angular/compiler-cli": "4.4.4",
        "@angular/core": "4.4.4",
        "@angular/forms": "4.4.4",
        "@angular/http": "4.4.4",
        "@angular/platform-browser": "4.4.4",
        "@angular/platform-browser-dynamic": "4.4.4",
        "@angular/platform-server": "4.4.4",
        "@ionic-native/app-version": "^4.2.1",
        "@ionic-native/core": "^4.2.1",
        "@ionic-native/diagnostic": "^4.2.1",
        "@ionic-native/geolocation": "^4.2.1",
        "@ionic-native/in-app-browser": "^4.2.1",
        "@ionic-native/keyboard": "^4.2.1",
        "@ionic/cloud-angular": "^0.12.0",
        "@ionic/storage": "^2.0.1",
        "@ng-idle/core": "^2.0.0-beta.10",
        "@ng-idle/keepalive": "^2.0.0-beta.10",
        "angular-calendar": "^0.21.2",
        "angular2-moment": "^1.3.3",
        "angular2-number-picker": "^0.8.8",
        "cordova-android": "^6.3.0",
        "cordova-android-play-services-gradle-release": "^1.1.1",
        "cordova-android-support-gradle-release": "^1.1.1",
        "cordova-browser": "^5.0.0",
        "cordova-ios": "^4.4.0",
        "cordova-open-native-settings": "^1.4.1",
        "cordova-plugin-app-version": "^0.1.9",
        "cordova-plugin-camera": "^2.4.1",
        "cordova-plugin-compat": "^1.1.0",
        "cordova-plugin-crosswalk-certificate": "^2.0.1",
        "cordova-plugin-crosswalk-webview": "^2.3.0",
        "cordova-plugin-device": "^1.1.6",
        "cordova-plugin-file": "^4.3.3",
        "cordova-plugin-file-transfer": "^1.6.3",
        "cordova-plugin-filepath": "^1.0.2",
        "cordova-plugin-geolocation": "^2.4.3",
        "cordova-plugin-inappbrowser": "^1.7.1",
        "cordova-plugin-splashscreen": "^4.0.3",
        "cordova-plugin-statusbar": "^2.2.3",
        "cordova-plugin-whitelist": "^1.3.2",
        "cordova.plugins.diagnostic": "^3.6.7",
        "intl": "^1.2.5",
        "ionic-angular": "^3.6.1",
        "ionic-native": "^2.9.0",
        "ionic-plugin-keyboard": "^2.2.1",
        "ionic2-calendar": "^0.3.6",
        "ionicons": "3.0.0",
        "jquery": "^3.2.1",
        "preboot": "^5.0.0",
        "rxjs": "5.4.3",
        "sw-toolbox": "3.6.0",
        "zone.js": "^0.8.4"
    },
    "devDependencies": {
        "@angular/cli": "^1.0.4",
        "@ionic/app-scripts": "3.0.0",
        "cordova": "^7.0.1",
        "ionic": "3.12.0",
        "ts-node": "^3.3.0",
        "typescript": "2.4.0",
        "typings": "^2.1.1"
    },
    "cordovaPlugins": [
        "cordova-plugin-whitelist",
        "cordova-plugin-console",
        "cordova-plugin-statusbar",
        "cordova-plugin-device",
        "ionic-plugin-keyboard",
        "cordova-plugin-splashscreen"
    ],

"cordova": {
        "plugins": {
            "ionic-plugin-keyboard": {},
            "cordova-plugin-inappbrowser": {},
            "cordova-plugin-device": {},
            "cordova-plugin-splashscreen": {},
            "cordova-plugin-statusbar": {},
            "cordova-plugin-whitelist": {},
            "cordova-plugin-file": {},
      ,
            "cordova-plugin-file-transfer": {},
            "cordova-plugin-filepath": {},
            "cordova-plugin-compat": {},
            "cordova-plugin-geolocation": {
                "GEOLOCATION_USAGE_DESCRIPTION": " "
            },
            "cordova-plugin-crosswalk-certificate": {},
            "cordova-plugin-app-version": {},
            "cordova-plugin-crosswalk-webview": {
                "XWALK_VERSION": "23+",
                "XWALK_LITEVERSION": "xwalk_core_library_canary:17+",
                "XWALK_COMMANDLINE": "--disable-pull-to-refresh-effect",
                "XWALK_MODE": "embedded",
                "XWALK_MULTIPLEAPK": "true"
            },
            "cordova-open-native-settings": {},
            "cordova.plugins.diagnostic": {},
            "cordova-android-support-gradle-release": {
                "ANDROID_SUPPORT_VERSION": "25.+"
            },
            "cordova-android-play-services-gradle-release": {}
        },
        "platforms": [
            "android",
            "ios"
        ]
    }
```

**Comment by kensodemann** [comment]
Probably not an issue, but there is an errant comma on the line right after `"cordova-plugin-file": {},`

is that line supposed to contain something?

**Comment by jordanpurinton** [comment]
Yeah, sorry that was my bad. I removed some stuff to shorten it. It is correct on my end.

**Comment by jordanpurinton** [comment]
Tried removing and adding platform and I got this:

```
C:\.....>ionic cordova platform add android
> cordova platform add android --save
× Running command - failed!
[ERROR] Exception: Using cordova-fetch for cordova-android@~6.2.2

        Adding android project...
        Creating Cordova project for the Android platform:
                Path: platforms\android
                Package: XXX
                Name: XXX
                Activity: MainActivity
                Android target: android-25
        Subproject Path: CordovaLib
        Android project created with cordova-android@6.2.3
        Installing "cordova-android-play-services-gradle-release" for android
        Subproject Path: CordovaLib
        Installing "cordova-android-support-gradle-release" for android
        Error: Cannot read property 'length' of undefined
```

**Comment by kensodemann** [comment]
I couldn't install all of that due to dependency issues. Specifically, `cordova-android-play-services-gradle-release` refused to install.

Ignoring that, though, it ran fine on my machine. Perhaps the issue is with that plugin itself?

I also noticed you are using Crosswalk, which is deprecated. May want to look into removing that from your project if possible and in making sure your other dependencies all match what we release with our starter (I noticed some of the Angular ones did not, same with TypeScript I _think_). I doubt any of that is causing your issue, but could lead to other issues.

For your actual issue, you probably should dig through the `plugins\cordova-android-support-gradle-release\scripts\after-prepare.js` file and try debugging it. It looks like there are only a couple places that `length` is used in there. Also, check issues in that project's repo and see if there is anything logged.

It _could_ be a Windows issue (I am on a Mac)

I am going to close this issue as we are well out of Ionic Framework land (actually, we left that land a while ago, but I wanted to try and see if there was something obvious going on here).

I urge that you ask this question on our [forum](https://forum.ionicframework.com) or on our [slack channel](https://ionicworldwide.herokuapp.com/). The wider audience there may have run into this issue.
